src/VecSim/algorithms/svs/svs_tiered.h

```diff
@@ -528,60 +533,71 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
-            std::shared_lock<std::shared_mutex> frontend_lock{this->flatIndexGuard};
+            std::shared_lock flat_lock{this->flatIndexGuard};
 
             auto flat_index = this->GetFlatIndex();
+            const auto frontend_index_size = this->frontendIndex->indexSize();
             const size_t dim = flat_index->getDim();
+            labels_to_move.reserve(frontend_index_size);
+            vectors_to_move.reserve(frontend_index_size * dim);
 
-            // Update snapshot to sync with current frontend index status could me changed
-            // since first call to TakeSnapshot()
-            TakeSnapshot(&to_delete, &to_add);
-
-            labels_to_add.reserve(to_add.size());
-            vectors_to_add.reserve(to_add.size() * dim);
-
-            labels_to_add.insert(labels_to_add.end(), to_add.begin(), to_add.end());
-            for (auto label : labels_to_add) {
-                if (this->frontendIndex->isLabelExists(label)) {
-                    const auto id = flat_index->getIdOfLabel(label);
-                    if (id != INVALID_ID) {
-                        auto data = flat_index->getDataByInternalId(id);
-                        vectors_to_add.insert(vectors_to_add.end(), data, data + dim);
-                    }
-                }
+            for (idType i = 0; i < frontend_index_size; i++) {
+                labels_to_move.push_back(flat_index->getVectorLabel(i));
+                auto data = flat_index->getDataByInternalId(i);
+                vectors_to_move.insert(vectors_to_move.end(), data, data + dim);
             }
+            // reset journal to the current frontend index state
+            swaps_journal.clear();
         } // release frontend index
 
         { // lock backend index for writing and add vectors there
-            std::scoped_lock lock(this->mainIndexGuard);
+            std::lock_guard lock(this->mainIndexGuard);
             auto svs_index = GetSVSIndex();
-            assert(labels_to_add.size() == vectors_to_add.size() / this->frontendIndex->getDim());
-            svs_index->addVectors(vectors_to_add.data(), labels_to_add.data(),
-                                  labels_to_add.size());
+            assert(labels_to_move.size() == vectors_to_move.size() / this->frontendIndex->getDim());
+            svs_index->addVectors(vectors_to_move.data(), labels_to_move.data(),
+                                  labels_to_move.size());
         }
         // clean-up frontend index
         { // lock frontend index for writing and delete moved vectors
-            std::scoped_lock lock(this->flatIndexGuard, this->journal_mutex);
-            // avoid deleting vectors from frontend which are updated/modified in the meantime
-            for (auto &p : this->journal) {
-                to_add.erase(p.first);
+            std::lock_guard lock(this->flatIndexGuard);
+
+            // Enumerate journal and swap label in labels_to_move to newId position.
```

Comment:
`swaps_journal` was cleared before we started to move vectors to the backend index. So here we are iterating the vector inserted into the flat index in the meantime, right?